src/ui/src/utils/analytics.ts

```diff
@@ -31,12 +36,103 @@ declare global {
 // Use this class instead of `window.analytics`.
 // This should be the only user of window.analytics.
 class PixieAnalytics {
+  waiting = false;
+
   loaded = false;
 
+  loggedIn = false;
+
   optOut = false;
 
+  /**
+   * Called automatically when PixieAnalytics is imported.
+   *
+   * Should be called manually whenever user logs in/out or opts in/out of analytics, to reset ID and permissions.
+   */
+  async reinit() {
+    // Due to a timing issue with lazy evaluation and how Webpack exports simple classes, we have to wait just a brief
+    // moment for PixieAPIManager to be defined. There is no cyclical dependency here, just a timing problem.
+    // This does not affect other consumers: all of them happen to wait (for React, or for history events, etc) already.
+    // This consumer, however, uses PixieAPIManager the moment this file is imported, and triggers the issue.
+    await Promise.resolve();
+    if (!PixieAPIManager) return; // Happens in unit tests for the same reason - just importing it causes issues.
+
+    try {
+      this.loggedIn = await PixieAPIManager.instance.isAuthenticated();
+    } catch {
+      this.loggedIn = false;
+    }
+
+    this.optOut = await this.checkOptOut();
+    if (!this.optOut) {
+      this.load();
+      await this.autoIdentify();
+    } else {
+      if (this.loaded) {
+        window.analytics.reset();
+      }
+    }
+  }
+
+  private async autoIdentify() {
+    if (!this.loggedIn) return;
+
+    const gqlClient = (PixieAPIManager.instance as PixieAPIClient).getCloudClient().graphQL;
+    try {
+      const { data: { user } } = await gqlClient.query<{
+        user: Pick<GQLUserInfo, 'id' | 'email' | 'orgName' >,
+      }>({
+        query: gql`
+          query getUserIdentityForAnalytics{
+            user {
+              id
+              email
+            }
+          }
+        `,
+      });
+
+      if (user?.id && user?.email) {
+        const allowDefaultLauncher = !isPixieEmbedded()
+          && ['/auth/login', '/auth/signup', '/login', '/signup'].includes(location.pathname);
+        window.analytics.identify(
+          user.id,
+          { email: user.email },
+          { integrations: { Intercom: { hideDefaultLauncher: !allowDefaultLauncher } } },
+        );
+      }
+    } catch {
+      // No-op: this happens when the user is not logged in; it's not an error.
+    }
+  }
+
+  private async checkOptOut() {
+    // To check if analytics are permitted to run, we test if there is a logged in user. If there is, check if they've
+    // opted out of analytics. If there isn't, they're permitted until that changes (for things like login/signup flow).
+    this.waiting = true;
```

Comment:
Should `this.waiting` be true even when we're waiting for `autoIdentify` to run?

Reply:
I think not, because `waiting` is about "waiting to learn if we have permission to run" rather than "waiting to make sure we're set up".

If we wanted to guarantee the `identify` call happens first in this sort of situation, we'd probably want to do something like buffering events that come in while `waiting` is true and dealing with the added complexity of that. However, we don't log events at a high enough frequency for that to be worth it I think.